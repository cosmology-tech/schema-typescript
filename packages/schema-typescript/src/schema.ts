import generate from "@babel/generator";
import * as t from "@babel/types";

import { SchemaTSContext, type SchemaTSOptions } from "./context";
import type { JSONSchema } from "./types";
import { getTypeNameSafe, isValidIdentifier, isValidIdentifierCamelized, makeCommentLine, shouldInclude, toCamelCase } from "./utils";

const getSchemaTypeNameSafe = (ctx: SchemaTSContext, str: string): string => {
  return getTypeNameSafe(ctx.options.namingStrategy, str);
};

const identifier = (name: string, typeAnnotation: t.TSTypeAnnotation) => {
  const i = t.identifier(name);
  i.typeAnnotation = typeAnnotation;
  return i;
};

const anyOrObjectWithUnknownProps = (ctx: SchemaTSContext) => {
  return ctx.options.strictTypeSafety ? t.tsTypeLiteral([
    t.tsIndexSignature(
      [
        identifier('key', t.tsTypeAnnotation(
          t.tsStringKeyword()
        ))
      ],
      t.tsTypeAnnotation(t.tsUnknownKeyword())
    )
  ]) : t.tsAnyKeyword();
}

export function generateTypeScriptTypes(schema: JSONSchema, options?: Partial<SchemaTSOptions>): t.ExportNamedDeclaration[] {
  const ctx = new SchemaTSContext(options, schema, schema, []);
  return generateInterfaces(ctx, schema);
}

export function generateTypeScript(schema: JSONSchema, options?: Partial<SchemaTSOptions>): string {
  const ctx = new SchemaTSContext(options, schema, schema, []);
  const interfaces = generateInterfaces(ctx, schema);
  return generate(t.file(t.program(interfaces))).code;
}

export function generateInterfaces(ctx: SchemaTSContext, schema: JSONSchema): t.ExportNamedDeclaration[] {
  const interfaces = [];

  try {
    // Process both $defs and definitions
    const definitions = schema.$defs || schema.definitions || {};
    for (const key in definitions) {
      if (shouldInclude(key, {
        include: ctx.options.include,
        exclude: ctx.options.exclude,
      })) {
        interfaces.push(createInterfaceDeclaration(ctx, key, definitions[key]));
      }
    }
  } catch (e) {
    console.error('Error processing interfaces');
    throw e;
  }

  // Process the main schema
  const title = schema.title;
  if (!title) {
    throw new Error('schema or options require a title');
  }
  if (shouldInclude(title, {
    include: ctx.options.include,
    exclude: ctx.options.exclude,
  })) {
    interfaces.push(createInterfaceDeclaration(ctx, title, schema));
  }
  return interfaces;
}

const createExportDeclarationForType = (
  ctx: SchemaTSContext,
  name: string,
  schema: JSONSchema,
  node: t.Declaration,
) => {
  const result =  t.exportNamedDeclaration(node);

  if (ctx.options.includeTypeComments && schema.description) {

    if (name.includes('.')) {
      // for complex names, let's add them for clarity/mapping
      result.leadingComments = [makeCommentLine(name)[0], makeCommentLine(schema.description)[0]]
    } else {
      result.leadingComments = makeCommentLine(schema.description)
    }

  }
  return result;
}

function createInterfaceDeclaration(
  ctx: SchemaTSContext,
  name: string,
  schema: JSONSchema
): t.ExportNamedDeclaration {
  // Handle standard properties if they exist
  let bodyElements: any = [];
  if (schema.properties) {
    const properties = schema.properties;
    const required = schema.required || [];
    bodyElements = Object.keys(properties).map(key => {
      const prop = properties[key];
      return createPropertySignature(ctx, key, prop, required, schema);
    });
  }

  // Handling additionalProperties if they exist
  if (schema.additionalProperties) {
    const additionalType = typeof schema.additionalProperties === 'boolean' ?
      t.tsStringKeyword() : getTypeForProp(ctx, schema.additionalProperties, [], schema);
    const indexSignature = t.tsIndexSignature(
      [t.identifier("key")], // index name, can be any valid name
      t.tsTypeAnnotation(additionalType)
    );
    indexSignature.parameters[0].typeAnnotation = t.tsTypeAnnotation(t.tsStringKeyword());
    bodyElements.push(indexSignature);
  }

  // Handling oneOf, anyOf, allOf if properties are not defined
  if (!schema.properties && (schema.oneOf || schema.anyOf || schema.allOf)) {
    const types = [];
    if (schema.oneOf) {
      types.push(getTypeForProp(ctx, { oneOf: schema.oneOf }, [], schema));
    }
    if (schema.anyOf) {
      types.push(getTypeForProp(ctx, { anyOf: schema.anyOf }, [], schema));
    }
    if (schema.allOf) {
      types.push(getTypeForProp(ctx, { allOf: schema.allOf }, [], schema));
    }

    // Create a union type if multiple types are generated
    const combinedType = types.length > 1 ? t.tsUnionType(types) : types[0];

    // Create a type alias instead of an interface if we're only handling these constructs
    const typeAlias = t.tsTypeAliasDeclaration(t.identifier(getSchemaTypeNameSafe(ctx, name)), null, combinedType);
    return createExportDeclarationForType(ctx, name, schema, typeAlias);
  }

  // Finally, create the interface declaration if there are any body elements
  if (bodyElements.length > 0) {
    const interfaceDeclaration = t.tsInterfaceDeclaration(
      t.identifier(getSchemaTypeNameSafe(ctx, name)),
      null,
      [],
      t.tsInterfaceBody(bodyElements)
    );
    return createExportDeclarationForType(ctx, name, schema, interfaceDeclaration);
  }

  if (schema.type) {
    return createExportDeclarationForType(ctx, name, schema, t.tsTypeAliasDeclaration(t.identifier(getSchemaTypeNameSafe(ctx, name)), null, getTypeForProp(ctx, schema, [], schema)))    
  }

  if (ctx.options.overrides && Object.prototype.hasOwnProperty.call(ctx.options.overrides, name)) {
    return createExportDeclarationForType(ctx, name, schema,
      t.tsTypeAliasDeclaration(t.identifier(getSchemaTypeNameSafe(ctx, name)), null,
      getTypeForProp(ctx, ctx.options.overrides[name], [], schema)
    ));
  }

  // Fallback to exporting a basic type if nothing else is possible
  // console.warn(`No properties or type definitions found for ${name}, defaulting to 'any'.`);
  return createExportDeclarationForType(ctx, name, schema, t.tsTypeAliasDeclaration(t.identifier(getSchemaTypeNameSafe(ctx, name)), null, t.tsAnyKeyword()))
}

function createPropertySignature(
  ctx: SchemaTSContext,
  key: string,
  prop: JSONSchema,
  required: string[],
  schema: JSONSchema
): t.TSPropertySignature {
  let camelCaseFn: (str: string) => string = toCamelCase;
  if (ctx.options.camelCaseFn) camelCaseFn = ctx.options.camelCaseFn;
  const name = ctx.options.camelCase ? camelCaseFn(key) : key;
  const propType = getTypeForProp(ctx, prop, required, schema);
  let identifier: t.Identifier | t.StringLiteral;
  let isIdent: boolean;
  if (ctx.options.camelCase) {
    isIdent = isValidIdentifierCamelized(key);
  } else {
    isIdent = isValidIdentifier(key);
  }
  identifier = isIdent ? t.identifier(name) : t.stringLiteral(key);
  const propSig = t.tsPropertySignature(
    identifier,
    t.tsTypeAnnotation(propType)
  );
  propSig.optional = !required.includes(key);
  if (ctx.options.includePropertyComments && prop.description) {
    propSig.leadingComments = makeCommentLine(prop.description);
  }

  return propSig;
}

function getTypeForProp(ctx: SchemaTSContext, prop: JSONSchema, required: string[], schema: JSONSchema): t.TSType {
  if (prop.$ref) {
    return resolveRefType(ctx, prop.$ref, schema);
  }

  if (prop.enum) {
    const enumType = prop.enum.map(enumValue => t.tsLiteralType(t.stringLiteral(enumValue)));
    return t.tsUnionType(enumType);
  }

  if (prop.const) {
    return t.tsLiteralType(t.stringLiteral(prop.const));
  }

  if (prop.type) {
    if (Array.isArray(prop.type)) {
      const arrayType = prop.type.map(type => getTypeForProp(ctx, { type, items: prop.items }, [], schema));
      return t.tsUnionType(arrayType);
    }

    switch (prop.type) {
      case 'string':
        return t.tsStringKeyword();
      case 'number':
      case 'integer':
        return t.tsNumberKeyword();
      case 'boolean':
        return t.tsBooleanKeyword();
      case 'null':
        return t.tsNullKeyword();
      case 'array':
        if (prop.items) {
          return t.tsArrayType(getTypeForProp(ctx, prop.items, required, schema));
        } else {
          throw new Error('Array items specification is missing');
        }
      case 'object':
        if (prop.properties) {
          const nestedProperties = prop.properties;
          const nestedRequired = prop.required || [];
          const typeElements = Object.keys(nestedProperties).map(nestedKey => {
            const nestedProp = nestedProperties[nestedKey];
            return createPropertySignature(ctx, nestedKey, nestedProp, nestedRequired, schema);
          });
          return t.tsTypeLiteral(typeElements);
        } else {
          // Handle dynamic properties with strict type safety option
          return anyOrObjectWithUnknownProps(ctx);
        }
      default:
        return t.tsAnyKeyword();
    }
  }

  // Handling oneOf, anyOf, allOf
  if (prop.anyOf) {
    const types = prop.anyOf.map((subProp) => getTypeForProp(ctx, subProp, required, schema));
    return t.tsUnionType(types);
  }
  if (prop.allOf) {
    const types = prop.allOf.map((subProp) => getTypeForProp(ctx, subProp, required, schema));
    return t.tsIntersectionType(types);
  }
  if (prop.oneOf) {
    const types = prop.oneOf.map((subProp) => getTypeForProp(ctx, subProp, required, schema));
    return t.tsUnionType(types);
  }

  // Fallback when no types are defined
  return t.tsAnyKeyword()

}

function getTypeReferenceFromSchema(ctx: SchemaTSContext, schema: JSONSchema, definitionName: string): t.TSType | null {
  if (definitionName) {

    // if (ctx.options.overrides && Object.prototype.hasOwnProperty.call(ctx.options.overrides, definitionName)) {
    //   return getTypeForProp(ctx, ctx.options.overrides[definitionName], [], schema);
    // }

    if (schema.$defs && schema.$defs[definitionName]) {
      return t.tsTypeReference(t.identifier(getSchemaTypeNameSafe(ctx, definitionName)));
    } else if (schema.definitions && schema.definitions[definitionName]) {
      return t.tsTypeReference(t.identifier(getSchemaTypeNameSafe(ctx, definitionName)));
    }
  }
  return null;  // Return null if no type reference is found
}


function resolveRefType(ctx: SchemaTSContext, ref: string, schema: JSONSchema): t.TSType {
  const path = ref.split('/');
  const definitionName = path.pop();

  // Try to resolve the type reference from the local schema
  const localTypeReference = getTypeReferenceFromSchema(ctx, schema, definitionName);
  if (localTypeReference) {
    return localTypeReference;
  }

  // Try to resolve the type reference from the root schema
  const rootTypeReference = getTypeReferenceFromSchema(ctx, ctx.root, definitionName);
  if (rootTypeReference) {
    return rootTypeReference;
  }

  // If no definitions are found in either schema, throw an error
  throw new Error(`Reference ${ref} not found in definitions or $defs.`);
}
