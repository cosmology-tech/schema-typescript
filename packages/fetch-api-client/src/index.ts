import fetch from 'isomorphic-fetch';

interface RequestOptions<Params> {
  hostname: string;
  path: string;
  headers?: { [key: string]: string };
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
  params?: Params;
  timeout?: number;
}

export interface APIClientOptions {
  restEndpoint: string;
}

export class APIClient {
  private baseUrl: string;
  private defaultTimeout: number = 10000; // 10 seconds

  constructor(options: APIClientOptions) {
    this.baseUrl = options.restEndpoint;
  }

  private buildUrl(endpoint: string, query?: { [key: string]: any }): string {
    const url = new URL(endpoint, this.baseUrl);
    if (query) {
      Object.keys(query).forEach(key => url.searchParams.append(key, query[key]));
    }
    return url.toString();
  }

  private async request<Resp>(options: RequestOptions<any>): Promise<Resp> {
    const { path, headers, method, params, timeout } = options;
    const url = this.buildUrl(path, method === 'GET' && params);
    const fetchOptions: RequestInit = {
      method,
      headers,
      body: method !== 'GET' && method !== 'DELETE' ? JSON.stringify(params) : null,
    };

    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), timeout || this.defaultTimeout);
    fetchOptions.signal = controller.signal;

    const response = await fetch(url, fetchOptions);
    clearTimeout(id);

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    return response.json() as Promise<Resp>;
  }

  get<Resp = unknown>(endpoint: string, query?: { [key: string]: any }, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'GET',
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
      params: query, // Correctly use query parameters for GET
      hostname: '',
    });
  }

  post<Resp = unknown, Params = any>(endpoint: string, params: Params, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...opts.headers,
      },
      params,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }

  put<Resp = unknown, Params = any>(endpoint: string, params: Params, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        ...opts.headers,
      },
      params,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }

  patch<Resp = unknown, Params = any>(endpoint: string, params: Params, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        ...opts.headers,
      },
      params,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }

  delete<Resp = unknown>(endpoint: string, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'DELETE',
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }
}
