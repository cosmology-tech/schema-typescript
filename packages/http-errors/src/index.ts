export class HTTPError extends Error {
  constructor(message: string, public status: number) {
    super(`${message} | Status: ${status}`);
    this.name = this.constructor.name;
  }
}
    
export class BadRequestError extends HTTPError {
  constructor() { super('Bad Request', 400); }
}
export class UnauthorizedError extends HTTPError {
  constructor() { super('Unauthorized', 401); }
}
export class PaymentRequiredError extends HTTPError {
  constructor() { super('Payment Required', 402); }
}
export class ForbiddenError extends HTTPError {
  constructor() { super('Forbidden', 403); }
}
export class NotFoundError extends HTTPError {
  constructor() { super('Not Found', 404); }
}
export class MethodNotAllowedError extends HTTPError {
  constructor() { super('Method Not Allowed', 405); }
}
export class NotAcceptableError extends HTTPError {
  constructor() { super('Not Acceptable', 406); }
}
export class ProxyAuthenticationRequiredError extends HTTPError {
  constructor() { super('Proxy Authentication Required', 407); }
}
export class RequestTimeoutError extends HTTPError {
  constructor() { super('Request Timeout', 408); }
}
export class ConflictError extends HTTPError {
  constructor() { super('Conflict', 409); }
}
export class GoneError extends HTTPError {
  constructor() { super('Gone', 410); }
}
export class LengthRequiredError extends HTTPError {
  constructor() { super('Length Required', 411); }
}
export class PreconditionFailedError extends HTTPError {
  constructor() { super('Precondition Failed', 412); }
}
export class PayloadTooLargeError extends HTTPError {
  constructor() { super('Payload Too Large', 413); }
}
export class URITooLongError extends HTTPError {
  constructor() { super('URI Too Long', 414); }
}
export class UnsupportedMediaTypeError extends HTTPError {
  constructor() { super('Unsupported Media Type', 415); }
}
export class RangeNotSatisfiableError extends HTTPError {
  constructor() { super('Range Not Satisfiable', 416); }
}
export class ExpectationFailedError extends HTTPError {
  constructor() { super('Expectation Failed', 417); }
}
export class ImATeapotError extends HTTPError {
  constructor() { super("I'm a Teapot", 418); }
}
export class MisdirectedRequestError extends HTTPError {
  constructor() { super('Misdirected Request', 421); }
}
export class UnprocessableEntityError extends HTTPError {
  constructor() { super('Unprocessable Entity', 422); }
}
export class LockedError extends HTTPError {
  constructor() { super('Locked', 423); }
}
export class FailedDependencyError extends HTTPError {
  constructor() { super('Failed Dependency', 424); }
}
export class UpgradeRequiredError extends HTTPError {
  constructor() { super('Upgrade Required', 426); }
}
export class PreconditionRequiredError extends HTTPError {
  constructor() { super('Precondition Required', 428); }
}
export class TooManyRequestsError extends HTTPError {
  constructor() { super('Too Many Requests', 429); }
}
export class RequestHeaderFieldsTooLargeError extends HTTPError {
  constructor() { super('Request Header Fields Too Large', 431); }
}
export class UnavailableForLegalReasonsError extends HTTPError {
  constructor() { super('Unavailable For Legal Reasons', 451); }
}
export class InternalServerError extends HTTPError {
  constructor() { super('Internal Server Error', 500); }
}
export class NotImplementedError extends HTTPError {
  constructor() { super('Not Implemented', 501); }
}
export class BadGatewayError extends HTTPError {
  constructor() { super('Bad Gateway', 502); }
}
export class ServiceUnavailableError extends HTTPError {
  constructor() { super('Service Unavailable', 503); }
}
export class GatewayTimeoutError extends HTTPError {
  constructor() { super('Gateway Timeout', 504); }
}
export class HTTPVersionNotSupportedError extends HTTPError {
  constructor() { super('HTTP Version Not Supported', 505); }
}
export class VariantAlsoNegotiatesError extends HTTPError {
  constructor() { super('Variant Also Negotiates', 506); }
}
export class InsufficientStorageError extends HTTPError {
  constructor() { super('Insufficient Storage', 507); }
}
export class LoopDetectedError extends HTTPError {
  constructor() { super('Loop Detected', 508); }
}
export class NotExtendedError extends HTTPError {
  constructor() { super('Not Extended', 510); }
}
export class NetworkAuthenticationRequiredError extends HTTPError {
  constructor() { super('Network Authentication Required', 511); }
}
    
  
export function createHTTPError(status: number): HTTPError {
  switch (status) {
  case 400: return new BadRequestError();
  case 401: return new UnauthorizedError();
  case 402: return new PaymentRequiredError();
  case 403: return new ForbiddenError();
  case 404: return new NotFoundError();
  case 405: return new MethodNotAllowedError();
  case 406: return new NotAcceptableError();
  case 407: return new ProxyAuthenticationRequiredError();
  case 408: return new RequestTimeoutError();
  case 409: return new ConflictError();
  case 410: return new GoneError();
  case 411: return new LengthRequiredError();
  case 412: return new PreconditionFailedError();
  case 413: return new PayloadTooLargeError();
  case 414: return new URITooLongError();
  case 415: return new UnsupportedMediaTypeError();
  case 416: return new RangeNotSatisfiableError();
  case 417: return new ExpectationFailedError();
  case 418: return new ImATeapotError();
  case 421: return new MisdirectedRequestError();
  case 422: return new UnprocessableEntityError();
  case 423: return new LockedError();
  case 424: return new FailedDependencyError();
  case 426: return new UpgradeRequiredError();
  case 428: return new PreconditionRequiredError();
  case 429: return new TooManyRequestsError();
  case 431: return new RequestHeaderFieldsTooLargeError();
  case 451: return new UnavailableForLegalReasonsError();
  case 500: return new InternalServerError();
  case 501: return new NotImplementedError();
  case 502: return new BadGatewayError();
  case 503: return new ServiceUnavailableError();
  case 504: return new GatewayTimeoutError();
  case 505: return new HTTPVersionNotSupportedError();
  case 506: return new VariantAlsoNegotiatesError();
  case 507: return new InsufficientStorageError();
  case 508: return new LoopDetectedError();
  case 510: return new NotExtendedError();
  case 511: return new NetworkAuthenticationRequiredError();
  default: return new HTTPError('Unknown Error', status);
  }
}
  