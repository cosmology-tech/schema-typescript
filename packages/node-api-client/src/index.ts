import * as http from 'http';
import * as querystring from 'querystring';
import { URLSearchParams } from 'url';

interface RequestOptions<Params> {
  hostname: string;
  path: string;
  headers?: { [key: string]: string };
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
  params?: Params;
  timeout?: number;
}

export interface APIClientOptions {
  restEndpoint: string;
}

export interface APIClientRequestHeaders {
  accept?: string | string[] | undefined;
  'accept-charset'?: string | string[] | undefined;
  'accept-encoding'?: string | string[] | undefined;
  'accept-language'?: string | string[] | undefined;
  'accept-ranges'?: string | undefined;
  'access-control-allow-credentials'?: string | undefined;
  'access-control-allow-headers'?: string | undefined;
  'access-control-allow-methods'?: string | undefined;
  'access-control-allow-origin'?: string | undefined;
  'access-control-expose-headers'?: string | undefined;
  'access-control-max-age'?: string | undefined;
  'access-control-request-headers'?: string | undefined;
  'access-control-request-method'?: string | undefined;
  age?: string | undefined;
  allow?: string | undefined;
  authorization?: string | undefined;
  'cache-control'?: string | undefined;
  'cdn-cache-control'?: string | undefined;
  connection?: string | string[] | undefined;
  'content-disposition'?: string | undefined;
  'content-encoding'?: string | undefined;
  'content-language'?: string | undefined;
  'content-length'?: string | number | undefined;
  'content-location'?: string | undefined;
  'content-range'?: string | undefined;
  'content-security-policy'?: string | undefined;
  'content-security-policy-report-only'?: string | undefined;
  cookie?: string | string[] | undefined;
  dav?: string | string[] | undefined;
  dnt?: string | undefined;
  date?: string | undefined;
  etag?: string | undefined;
  expect?: string | undefined;
  expires?: string | undefined;
  forwarded?: string | undefined;
  from?: string | undefined;
  host?: string | undefined;
  'if-match'?: string | undefined;
  'if-modified-since'?: string | undefined;
  'if-none-match'?: string | undefined;
  'if-range'?: string | undefined;
  'if-unmodified-since'?: string | undefined;
  'last-modified'?: string | undefined;
  link?: string | string[] | undefined;
  location?: string | undefined;
  'max-forwards'?: string | undefined;
  origin?: string | undefined;
  prgama?: string | string[] | undefined;
  'proxy-authenticate'?: string | string[] | undefined;
  'proxy-authorization'?: string | undefined;
  'public-key-pins'?: string | undefined;
  'public-key-pins-report-only'?: string | undefined;
  range?: string | undefined;
  referer?: string | undefined;
  'referrer-policy'?: string | undefined;
  refresh?: string | undefined;
  'retry-after'?: string | undefined;
  'sec-websocket-accept'?: string | undefined;
  'sec-websocket-extensions'?: string | string[] | undefined;
  'sec-websocket-key'?: string | undefined;
  'sec-websocket-protocol'?: string | string[] | undefined;
  'sec-websocket-version'?: string | undefined;
  server?: string | undefined;
  'set-cookie'?: string | string[] | undefined;
  'strict-transport-security'?: string | undefined;
  te?: string | undefined;
  trailer?: string | undefined;
  'transfer-encoding'?: string | undefined;
  'user-agent'?: string | undefined;
  upgrade?: string | undefined;
  'upgrade-insecure-requests'?: string | undefined;
  vary?: string | undefined;
  via?: string | string[] | undefined;
  warning?: string | undefined;
  'www-authenticate'?: string | string[] | undefined;
  'x-content-type-options'?: string | undefined;
  'x-dns-prefetch-control'?: string | undefined;
  'x-frame-options'?: string | undefined;
  'x-xss-protection'?: string | undefined;
}
export type APIClientRequestOpts = {
  headers?: APIClientRequestHeaders;
  timeout?: number;
};
export class APIClient {
  private hostname: string;
  private port: number = 8001; // Default Kubernetes proxy port
  private defaultTimeout: number = 10000; // 10 seconds

  constructor(options: APIClientOptions) {
    const url = new URL(options.restEndpoint);
    this.hostname = url.hostname;
    this.port = url.port ? parseInt(url.port) : this.port;
  }

  private buildFullPath(endpoint: string, query?: { [key: string]: any }): string {
    const queryString = query ? `?${querystring.stringify(query)}` : '';
    return endpoint + queryString;
  }

  get<Resp = unknown>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: void,
    opts: APIClientRequestOpts = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    return this.request<Resp, void>({
      path: fullPath,
      method: 'GET',
      // @ts-ignore
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  post<Resp = unknown, Params = any>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: Params,
    opts: http.RequestOptions & { timeout?: number; isFormData?: boolean } = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    const headers: http.OutgoingHttpHeaders = opts.isFormData
      ? {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...opts.headers,
      }
      : {
        'Content-Type': 'application/json',
        ...opts.headers,
      };
    const bodyContent = opts.isFormData
      ? new URLSearchParams(body as any).toString()
      : JSON.stringify(body);
    headers['Content-Length'] = Buffer.byteLength(bodyContent).toString();

    return this.request<Resp, Params>({
      path: fullPath,
      method: 'POST',
      // @ts-ignore
      headers,
      // @ts-ignore
      params: bodyContent,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  patch<Resp = unknown, Params = any>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: Params,
    opts: http.RequestOptions & { timeout?: number; isFormData?: boolean } = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    const headers: http.OutgoingHttpHeaders = opts.isFormData
      ? {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...opts.headers,
      }
      : {
        'Content-Type': 'application/json',
        ...opts.headers,
      };
    const bodyContent = opts.isFormData
      ? new URLSearchParams(body as any).toString()
      : JSON.stringify(body);
    headers['Content-Length'] = Buffer.byteLength(bodyContent).toString();

    return this.request<Resp, Params>({
      path: fullPath,
      method: 'PATCH',
      // @ts-ignore
      headers,
      // @ts-ignore
      params: bodyContent,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  put<Resp = unknown, Params = any>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: Params,
    opts: http.RequestOptions & { timeout?: number } = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    const bodyContent = JSON.stringify(body);
    const headers = {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(bodyContent).toString(),
      ...opts.headers,
    };

    return this.request<Resp, Params>({
      path: fullPath,
      method: 'PUT',
      // @ts-ignore
      headers,
      // @ts-ignore
      params: bodyContent,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  delete<Resp = unknown, Params = any>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: void,
    opts: http.RequestOptions & { timeout?: number } = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    return this.request<Resp, void>({
      path: fullPath,
      method: 'DELETE',
      // @ts-ignore
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  private request<Resp, Params>(
    options: RequestOptions<Params>
  ): Promise<Resp> {
    return new Promise((resolve, reject) => {
      const requestOptions: http.RequestOptions = {
        hostname: this.hostname,
        port: this.port,
        path: options.path,
        method: options.method,
        headers: options.headers,
      };

      const req = http.request(requestOptions, (res) => {
        let data = '';
        res.on('data', (chunk) => {
          data += chunk;
        });
        res.on('end', () => {
          try {
            const parsedData: Resp = JSON.parse(data);
            resolve(parsedData);
          } catch (error) {
            reject(error);
          }
        });
      });

      req.on('error', (error) => {
        reject(error);
      });

      req.setTimeout(options.timeout, () => {
        req.abort();
        reject(new Error('Request timeout'));
      });

      if (options.params && ['POST', 'PUT', 'PATCH'].includes(options.method)) {
        req.write(options.params);
      }

      req.end();
    });
  }
}
