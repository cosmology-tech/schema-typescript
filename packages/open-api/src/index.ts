export * from './openapi';
export * from './openapi.types';
export * from './utils';