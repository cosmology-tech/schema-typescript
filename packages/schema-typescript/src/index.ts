export * from './context';
export * from './schema';
export * from './types';
export * from './utils';
