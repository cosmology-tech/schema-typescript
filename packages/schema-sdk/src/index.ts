import { getDefaultSchemaTSOptions } from 'schema-typescript';

export * from './openapi';
export * from './openapi.types';
export * from './utils';

export { getDefaultSchemaTSOptions };