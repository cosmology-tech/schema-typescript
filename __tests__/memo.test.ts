import { writeFileSync } from 'fs';

import schema from '../__fixtures__/schemas/memo.json';
import { generateTypeScript } from '../src';

it('memo', () => {
    const code = generateTypeScript(schema as any);
    expect(code).toMatchSnapshot();
    writeFileSync(__dirname + '/../__fixtures__/output/memo.ts', code);
});

it('memo camelCase', () => {
    const code = generateTypeScript(schema as any, {
        useSingleQuotes: true,
        camelCase: true
    });
    expect(code).toMatchSnapshot();
    writeFileSync(__dirname + '/../__fixtures__/output/memo.camel.ts', code);
});
