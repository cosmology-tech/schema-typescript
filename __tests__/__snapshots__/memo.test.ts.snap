// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`memo 1`] = `
"export interface Memo {
  $schema?: string;
  memo_keys: {
    key: string;
    description: string;
    git_repo: string;
    memo: any;
  }[];
}"
`;

exports[`memo camelCase 1`] = `
"export interface Memo {
  schema?: string;
  memoKeys: {
    key: string;
    description: string;
    gitRepo: string;
    memo: any;
  }[];
}"
`;
